Fix page table cursor invariants by exposing level details

junyang-zh opened the pull request:
This PR is based on #2595 and #2605 . Please review them first.

The problem is that for page table cursor's operations (`map`, `take_next`, and `protect_next`), they traverse the cursor forward to the next virtual address when the operations are finished. Assuming a cursor holding a range `0x1000..0x2000`, before `map` the cursor is at virtual address `0x1000`, and after mapping a page, the cursor is at `0x2000`. So the cursor can point to the end of the range.

But we want to maintain an invariant that the current virtual address must be within the range of each node in the path, so that every operation (like `query` or `map`) always succeeds. After the fix, the return values are drastically simplified.

The way to fix the problem is to expose level details to the user, and define `map`/`unmap`/`protect` to operate on the current PTE only, without a compulsory `find_next`.

---

This PR is also based on https://github.com/asterinas/asterinas/pull/2744 though not relevant.

junyang-zh committed:
Use generics to separate `Cursor` and `CursorMut`

junyang-zh committed:
No automatic level adjusting in `query`/`map`

lrh2000 commented on ostd/src/mm/vm_space.rs:
```diff
@@ -434,120 +472,91 @@ impl<'a> CursorMut<'a> {
+    pub fn unmap(&mut self) -> usize {
+        // SAFETY: It is safe to un-map memory in the userspace. And the
+        // un-mapped items are dropped after TLB flushes. So they outlive
+        // the current RCU lifetime `'a`.
```
I think you should try to resolve the merge conflicts correctly.

lrh2000 commented on ostd/src/mm/vm_space.rs:
```diff
@@ -434,120 +472,91 @@ impl<'a> CursorMut<'a> {
+                panic_guard.forget();
+                let rcu_frame = Frame::rcu_from_unsized(rcu_pt);
+                self.flusher
+                    .issue_tlb_flush_with(TlbFlushOp::for_range(va..va + len), rcu_frame);
```
Same? I'm not sure why you don't call `Frame::rcu_from_unsized`?

lrh2000 reviewed (commented):
I just checked a few random places in this PR. It's not a complete review. It's not even close to complete because I really only checked a few random places.

I'm submitting them anyway, since you should have an updated version of this PR after merging #2605.

junyang-zh committed:
Stop `Cursor` moving after operations

junyang-zh commented:
@boterinas codex

lrh2000 commented on ostd/src/mm/page_table/cursor/mod.rs:
```diff
@@ -459,31 +458,23 @@ impl<C: PageTableConfig> CursorMut<'_, C> {
-
-        frag
+    /// Panics if the current level is at the top level and the corresponding
+    /// [`PageTableConfig::TOP_LEVEL_CAN_UNMAP`] is false.
+    pub unsafe fn unmap(&mut self) -> Option<PageTableFrag<C>> {
+        if !C::TOP_LEVEL_CAN_UNMAP && self.level == C::NR_LEVELS {
+            panic!("Unmapping top-level page table nodes");
+        }
+        self.replace_cur_entry(PteState::Absent)
     }
```
### Make the cursor level externally observable and controllable.

Before this PR, the only externally observable and mutable state of a cursor was its current virtual address. I believe that all previous APIs will behave identically, regardless of the internal state of the cursor (e.g., the current level), as long as the virtual addresses are the same.

One big problem of the changes in this PR is that they will allow quite a few invalid states like:
```
barrier_va = HUGE_PAGE_SIZE - PAGE_SIZE..HUGE_PAGE_SIZE + PAGE_SIZE;
va = HUGE_PAGE_SIZE - PAGE_SIZE; // Or, `va == HUGE_PAGE_SIZE`.
level = 2;
// `protect()` and `unmap()` will succeed?
```

To me, those states appear to be much "more" invalid than the `va == barrier_va.end` case. But they are allowed in this PR. Also I don't think there is a way to forbid them.

So my current feelings are that making the cursor level externally observable is not ideal.

Additionally, it will make the code uglier than before. For example,
```rust
while cursor.push_if_exists().is_some() {}
match cursor.query() { /* .. */ }
```

But, on the other hand, the code changes in the PR make me wonder if we can have something like:
```rust
impl Cursor {
    fn query_leaf(&mut self) -> Entry<Leaf>;
    fn query_subtree(&mut self, level: PagingLevel) -> Entry<Subtree>;

    fn find_leaf(&mut self) -> Option<Leaf>;
    fn find_subtree(&mut self) -> Option<Subtree>;
}

enum Entry<T> {
    Value(T),
    Vacant(VacantEntry),
}

/// A cursor that stops at a PTE that containing an item.
struct Leaf {
    cursor: &mut Cursor,
    item: C::ItemRef,
}

impl Leaf {
    fn item(&self) -> C::ItemRef;

    fn protect(self, op: impl FnOnce());

    fn unmap(self);
}

/// A cursor that stops at a PTE that containing an item or points to a subtree.
struct Subtree {
    cursor: &mut Cursor,
    item: Option<C::ItemRef>, // Is it necessary?
}

impl Subtree {
    fn unmap(self);
}

/// A cursor that stops at a PTE that is empty.
struct VacantEntry {
    cursor: &mut Cursor,
}

impl VacantEntry {
    fn map(self, item: C::Item);
}
```

I think the above APIs are more reasonable, but I don't yet know how they interact with `VmSpace`'s public APIs.

lrh2000 commented on ostd/src/mm/page_table/cursor/mod.rs:
```diff
@@ -275,30 +280,26 @@ impl<'rcu, C: PageTableConfig, const MUTABLE: bool> Cursor_<'rcu, C, MUTABLE> {
         }
 
-        // FIXME: Maintain the `self.barrier_va.contains(self.va)` invariant:
-        // <https://github.com/asterinas/asterinas/pull/2613>.
```
### Forbid `self.va == barrier_va`.

What is the exact benefit that this type invariant provides?

Before:
```rust
while cursor.find_next().is_some() {
    cursor.unmap_next();
}
```
After:
```rust
while cursor.find_next().is_some() {
    cursor.unmap_next();
    if cursor.jump_next().is_err() {
        break;
    }
}
```

I think the code looks uglier this way. Using `self.va == barrier_va.end` to indicate that the cursor has been exhausted does _not_ seem like a bad idea.

From the PR description:
> But we want to maintain an invariant that the current virtual address must be within the range of each node in the path, so that every operation (like query or map) always succeeds.

It is impossible to ensure that `map()` always succeeds because mapping a page that was already mapped (maybe even at a different level) or mapping a too-large page must fail. We cannot know if or when this will happen at compile time.

For `protect()`/`unmap()`, refer to my comment above. The current approach to make it always succeed is not correct.

I agree that there is a benefit to having `query()` always succeed, but it seems minimal. It does not seem to outweigh the disadvantages of the ugly code related to the use cases of `find_next()`.

lrh2000 commented on ostd/src/mm/vm_space.rs:
```diff
@@ -306,6 +302,16 @@ impl Cursor<'_> {
+    /// Get the current virtual address range of the cursor.
+    pub fn cur_va_range(&self) -> Range<Vaddr> {
+        self.0.cur_va_range()
+    }
```
`Get` -> `Gets`

lrh2000 commented on ostd/src/mm/page_table/cursor/mod.rs:
```diff
@@ -275,30 +280,26 @@ impl<'rcu, C: PageTableConfig, const MUTABLE: bool> Cursor_<'rcu, C, MUTABLE> {
         None
```
You should remove the code and convert the above `while` loop to `loop { ... return None; }`.

`while self.va < end` is always true.

lrh2000 commented on ostd/src/mm/page_table/cursor/mod.rs:
```diff
@@ -459,31 +458,23 @@ impl<C: PageTableConfig> CursorMut<'_, C> {
+    /// [`PageTableConfig::TOP_LEVEL_CAN_UNMAP`] is false.
+    pub unsafe fn unmap(&mut self) -> Option<PageTableFrag<C>> {
+        if !C::TOP_LEVEL_CAN_UNMAP && self.level == C::NR_LEVELS {
+            panic!("Unmapping top-level page table nodes");
```
```suggestion
            panic!("unmapping top-level page table nodes");
```

lrh2000 commented on ostd/src/mm/page_table/cursor/mod.rs:
```diff
@@ -349,35 +350,68 @@ impl<'rcu, C: PageTableConfig, const MUTABLE: bool> Cursor_<'rcu, C, MUTABLE> {
     ///
     /// # Panics
     ///
     /// This function will panic if
```
```suggestion
    /// Panics if
```

This seems the only exception in the whole file. All other doc comments use "Panics if"?

lrh2000 commented on ostd/src/mm/page_table/node/entry.rs:
```diff
@@ -41,9 +41,10 @@ impl<'a, 'rcu, C: PageTableConfig> Entry<'a, 'rcu, C> {
+        //  - The PTE was read from the node, which contains valid PTEs;
+        //  - the child pointed to by the PTE outlives the reference, since
+        //    either PTs and mapped items outlive `'rcu`;
+        //  - the level matches the current node.
```
I think we use:
```rust
// The caller must ensure that:   // <- This is an incomplete sentence.
// - aaa;
// - aaa;
// - aaa.
```
And:
```rust
// SAFETY:  // <- This is not a sentence.
// - Aaa.
// - Aaa.
// - Aaa.
```

However, it seems that you use a style that combines the two above. I think we should use the second style instead.

lrh2000 commented on ostd/src/mm/page_table/node/mod.rs:
```diff
@@ -187,6 +187,17 @@ impl<'rcu, C: PageTableConfig> PageTableGuard<'rcu, C> {
+        //  - The PTE must be the result of `PteState::into_pte` since it was
+        //    written into the node;
+        //  - the child outlives `'rcu`;
+        //  - the level matches the residing node.
```
Same here.

lrh2000 commented on ostd/src/mm/page_table/node/entry.rs:
```diff
@@ -209,8 +210,7 @@ impl<'a, 'rcu, C: PageTableConfig> Entry<'a, 'rcu, C> {
-        // SAFETY: The index is within the bound.
-        let pte = unsafe { guard.read_pte(idx) };
+        let pte = guard.read_pte(idx);
         Self {
```
The safety comment is outdated.

lrh2000 commented on ostd/src/mm/page_table/cursor/mod.rs:
```diff
@@ -48,13 +48,29 @@ use crate::{
+/// types aliases.
 #[derive(Debug)]
-pub(crate) struct Cursor<'rcu, C: PageTableConfig> {
+pub(crate) struct Cursor_<'rcu, C: PageTableConfig, const MUTABLE: bool = false> {
```
I was thinking about to use:
```rust
impl Deref for CursorMut {
    fn deref(&self) -> &Cursor;
}
```
which is guaranteed not to monomorphized the same implementation twice.

---

> ffffffff883657f0 l     F .text  000000000000011f .hidden _RNvMNtNtNtCs8LA18vg2m3n_4ostd2mm10page_table6cursorINtB2_7Cursor_NtNtB6_8vm_space12UserPtConfigE20push_level_if_existsB8_
> ffffffff883657f0 l     F .text  000000000000011f .hidden _RNvMNtNtNtCs8LA18vg2m3n_4ostd2mm10page_table6cursorINtB2_7Cursor_NtNtB6_8vm_space12UserPtConfigKb1_E20push_level_if_existsB8_

Perhaps the const generic approach is also fine if the release build will merge two identical implementations.There are two symbols but they're both referring to the code at `ffffffff883657f0`.

I also tried to figure out a better name than `Cursor_`. But I failed.

lrh2000 reviewed (commented):
Sorry for the extremely long delay...

So there are two major changes in this PR;
 - Make the cursor level externally observable and controllable.
 - Forbid `self.va == barrier_va`.

To be honest, checking the changes in this PR, I am not entirely sure either of them look the right direction. For details, see the comments below.

Note that maybe your part of motivation is to enable CortenMM, however, I did not (and cannot, due to lack of knowledge) take that into consideration. So if it's part of your motivation you should elaborate more.

junyang-zh commented on kernel/src/vm/vmar/vmar_impls/remap.rs:
```diff
@@ -199,31 +202,30 @@ impl Vmar {
                     let frame = (*frame).clone();
 
-                    cursor.unmap(PAGE_SIZE);
+                    cursor.unmap();
```
No need. Currently userspace cannot map huge pages so it is so far good.

junyang-zh commented:
> Note that maybe your part of motivation is to enable CortenMM, however, I did not (and cannot, due to lack of knowledge) take that into consideration. So if it's part of your motivation you should elaborate more.

Thank you. 

> Make the cursor level externally observable and controllable.

The motivation to make the cursor level externally visible is to enable CortenMM. Since when storing extra state (VMA pieces actually) in the page table, the caller need to put it in the correct level. Therefore the cursor must allow the user to navigate to specific levels and do operations.

> Forbid `self.va == barrier_va`.

This is not actually related with CortenMM. It's because I find the by-default advance behavior of `map`, `unmap` and `protect` not very tidy at some places, e.g., mapping or skipping already present pages:

https://github.com/asterinas/asterinas/blob/6c314ef6da150a9ce394642bbb0d89c81b045da6/kernel/src/vm/vmar/vm_mapping.rs#L621-L628

, some inconsistencies:

https://github.com/asterinas/asterinas/blob/6c314ef6da150a9ce394642bbb0d89c81b045da6/kernel/src/vm/vmar/vmar_impls/fork.rs#L105-L108

, and jumping back after `unmap` to `map`:

https://github.com/asterinas/asterinas/blob/6c314ef6da150a9ce394642bbb0d89c81b045da6/kernel/src/vm/vmar/vm_mapping.rs#L474-L476

If we do not let `map`, `unmap` and `protect` advance after operation by default, there seems no need to have `self.va == barrier_va`. Therefore I disallowed it. Another possible approach would be not advance by default and allow `self.va == barrier_va`. So the only case that will exhaust the cursor is `find_next`. It seems to resolve your concerns about the tidiness of `find_next` cases.

lrh2000 commented:
> The motivation to make the cursor level externally visible is to enable CortenMM. Since when storing extra state (VMA pieces actually) in the page table, the caller need to put it in the correct level. Therefore the cursor must allow the user to navigate to specific levels and do operations.

OK. Then will the APIs listed in https://github.com/asterinas/asterinas/pull/2613#discussion_r3649609756 work for you?

One problem I am aware of is that: There will be a disaster if we have all the structures and methods duplicate for the `VmSpace` cursor APIs.

If it's the only problem, I hope we can find a way to resolve it. I'm not sure how to do that at the moment, though.

> mapping or skipping already present pages:
> , some inconsistencies:
> , and jumping back after `unmap` to `map`:

Thanks for the explanation! Avoiding the automatic advance of the cursor can simplify some code. That makes sense.

It seems that both have their use cases, in an ideal world, maybe we can end up with something like:
```
cursor.find().map().next();
```
This allows each caller to decide whether to move forward.

But perhaps this is a fairly minor point.... and we don't really need to do it if implementing so is difficult.

> So the only case that will exhaust the cursor is `find_next`. It seems to resolve your concerns about the tidiness of `find_next` cases.

I guess this is not very related...
```patch
 while cursor.find_next().is_some() {
     cursor.unmap_next();
+    if cursor.jump_next().is_err() {
+        break;
+    }
 }
```

I want to use this use case to argue that `unmap` should advance the cursor. However, you also provided me with some counterexamples that suggest advancing might be redundant. What you said also makes sense to me.

junyang-zh commented:
> Then will the APIs listed in https://github.com/asterinas/asterinas/pull/2613#discussion_r3649609756 work for you?

It is indeed a good suggestion and I can understand. I perhaps need some time to see if it requires too much effort.

junyang-zh commented on kernel/core/src/vm/vmar/vm_mapping.rs:
```diff
@@ -481,15 +483,15 @@ impl VmMapping {
                     }
                     cursor.flusher().sync_tlb_flush();
                 }
```
Yes but not related with this branch. Fix in https://github.com/asterinas/asterinas/pull/3651

junyang-zh committed:
Replace `Cursor` API's `len` specified range style to `end`

This is to improve the use case in `while cursor.find_next(end)` loops. Ends
are always fixed, while len needs to be calculated in the condition.

junyang-zh commented:
I prototyped the proposal of @lrh2000 at https://github.com/asterinas/asterinas/commit/93352fd7009c8754eb5dff8762ba3aef4ca0aebb, with some help from AI tools. It can make some call site clearer and make invalid state explicit.

My concern of this approach is that it may introduce much complexities into OSTD (around 400 diff lines), especially when we need to duplicate it into VmSpace cursor. I currently do not have an idea to solve it. The benefited caller sites are not many. Therefore I still prefer the current solution.

I think this state machine style API can be good if we can make it a safe wrapper for current raw OSTD APIs. Then we can balance OSTD complexity and the caller's ease.

Further thoughts @lrh2000 ? Do the proposal at https://github.com/asterinas/asterinas/commit/93352fd7009c8754eb5dff8762ba3aef4ca0aebb implement what you think?

lrh2000 commented:
What is your response to the "problem" part? My main motivation is to forbid invalid states at the type level. Assuming we keep the current implementation, how can we best resolve this issue? Just add more assertions?

The "problem" part is:
> One big problem of the changes in this PR is that they will allow quite a few invalid states like:
> 
> ```
> barrier_va = HUGE_PAGE_SIZE - PAGE_SIZE..HUGE_PAGE_SIZE + PAGE_SIZE;
> va = HUGE_PAGE_SIZE - PAGE_SIZE; // Or, `va == HUGE_PAGE_SIZE`.
> level = 2;
> // `protect()` and `unmap()` will succeed?
> ```
> 
> To me, those states appear to be much "more" invalid than the `va == barrier_va.end` case. But they are allowed in this PR. Also I don't think there is a way to forbid them.
> 
> So my current feelings are that making the cursor level externally observable is not ideal.

---

> My concern of this approach is that it may introduce much complexities into OSTD (around 400 diff lines), especially when we need to duplicate it into VmSpace cursor. I currently do not have an idea to solve it.

Could you analyze it further to determine where it comes from?

From what I can see:
 1. I don't understand why the `VmSpace` types do not wrap the `Subtree`/`Leaf` types. Instead, they wrap `CursorMut`. This leads to duplicate checks at runtime and bloats the code with quite a few `unreachable!()`.
 2. For some reason, AI is generally "good" at defensive programming, so it inserts code like this:
    ```
                        let level = self
                            .0
                            .level
                            .checked_sub(1)
                            .expect("a base-page mapping cannot cross the cursor boundary");
    ```
    Code like this looks awful and makes it longer.
 3. We can avoid defining and implementing `unmap()` twice (once for `Subtree` and once for `Leaf`) by introducing `Leaf::to_subtree() -> Subtree`.
 4. Part of the code is boilerplate code, such as `impl Deref{,Mut}`. If we can have `#[derive(Deref, DerefMut)]`, they may go away.
 5. ...

> The benefited caller sites are not many.

It seems that there are quite `unreachable!()` which comes from `entry.unmap()` + `cursor.query()`. I think `entry.unmap()` should probably return a `VacantEntry`. So the code will look much better.

junyang-zh commented:
> What is your response to the "problem" part?

Sorry I missed that. However just add more assertions seems OK. Nearly no additional code since `map` already checks them, so we just need to add the same checks to `protect` and `unmap`. `query` can still have the current range exceeding the barrier.

<details><summary>Diff</summary>
<p>

```
diff --git a/ostd/src/mm/page_table/cursor/mod.rs b/ostd/src/mm/page_table/cursor/mod.rs
index e01146506..bac316066 100644
--- a/ostd/src/mm/page_table/cursor/mod.rs
+++ b/ostd/src/mm/page_table/cursor/mod.rs
@@ -429,7 +429,8 @@ impl<'rcu, C: PageTableConfig> CursorMut<'rcu, C> {
     ///  - the cursor level does not match the level of the item to be mapped;
     ///  - the current virtual address is not aligned to the page size of the
     ///    item to be mapped;
-    ///  - the end of the current virtual address range exceeds the locked range;
+    ///  - the current virtual address range is not fully contained in the
+    ///    cursor range;
     ///  - the current virtual address range contains mappings.
     ///
     /// # Safety
@@ -438,8 +439,6 @@ impl<'rcu, C: PageTableConfig> CursorMut<'rcu, C> {
     ///  - the range being mapped does not affect kernel's memory safety;
     ///  - the physical address to be mapped is valid and safe to use.
     pub unsafe fn map(&mut self, item: C::Item) {
-        debug_assert!(self.va < self.barrier_va.end);
-
         let (_, level, _) = C::item_raw_info(&item);
         assert!(
             level <= C::HIGHEST_TRANSLATION_LEVEL,
@@ -449,17 +448,7 @@ impl<'rcu, C: PageTableConfig> CursorMut<'rcu, C> {
             self.level, level,
             "cursor level do not match the item mapping level"
         );
-        let size = page_size::<C>(level);
-        assert_eq!(
-            self.va % size,
-            0,
-            "cursor virtual address not aligned for mapping"
-        );
-        let end = self.va + size;
-        assert!(
-            end <= self.barrier_va.end,
-            "cursor virtual address out-of-bound for mapping"
-        );
+        self.assert_cur_va_range_valid();
 
         if !matches!(self.cur_entry().to_ref(), PteStateRef::Absent) {
             panic!("mapping over an already mapped page");
@@ -488,9 +477,15 @@ impl<'rcu, C: PageTableConfig> CursorMut<'rcu, C> {
     ///
     /// # Panics
     ///
-    /// Panics if the current level is at the top level and the corresponding
-    /// [`PageTableConfig::TOP_LEVEL_CAN_UNMAP`] is false.
+    /// Panics if:
+    ///  - the current virtual address is not aligned to the current entry
+    ///    range;
+    ///  - the current virtual address range is not fully contained in the
+    ///    cursor range;
+    ///  - the current level is at the top level and the corresponding
+    ///    [`PageTableConfig::TOP_LEVEL_CAN_UNMAP`] is false.
     pub unsafe fn unmap(&mut self) -> Option<PageTableFrag<C>> {
+        self.assert_cur_va_range_valid();
         if !C::TOP_LEVEL_CAN_UNMAP && self.level == C::NR_LEVELS {
             panic!("unmapping top-level page table nodes");
         }
@@ -507,6 +502,14 @@ impl<'rcu, C: PageTableConfig> CursorMut<'rcu, C> {
     /// It only modifies the page properties if the current entry state is
     /// [`PteState::Mapped`]. Otherwise, it does nothing.
     ///
+    /// # Panics
+    ///
+    /// Panics if:
+    ///  - the current virtual address is not aligned to the current entry
+    ///    range;
+    ///  - the current virtual address range is not fully contained in the
+    ///    cursor range.
+    ///
     /// # Safety
     ///
     /// The caller should ensure that:
@@ -515,9 +518,22 @@ impl<'rcu, C: PageTableConfig> CursorMut<'rcu, C> {
     ///  - the privileged flag `AVAIL1` should not be altered, since this flag
     ///    is reserved for all page tables.
     pub unsafe fn protect(&mut self, op: &mut impl FnMut(&mut PageProperty)) {
+        self.assert_cur_va_range_valid();
         self.cur_entry().protect(op);
     }
 
+    fn assert_cur_va_range_valid(&self) {
+        let cur_va_range = self.cur_va_range();
+        assert_eq!(
+            self.va, cur_va_range.start,
+            "current virtual address is not aligned to the current entry range"
+        );
+        assert!(
+            self.barrier_va.start <= cur_va_range.start && cur_va_range.end <= self.barrier_va.end,
+            "current virtual address range exceeds cursor range"
+        );
+    }
+
     fn replace_cur_entry(&mut self, new_child: PteState<C>) -> Option<PageTableFrag<C>> {
         let rcu_guard = self.rcu_guard;
 
diff --git a/ostd/src/mm/page_table/test.rs b/ostd/src/mm/page_table/test.rs
index c0877fb99..79bd05cfa 100644
--- a/ostd/src/mm/page_table/test.rs
+++ b/ostd/src/mm/page_table/test.rs
@@ -471,7 +471,7 @@ mod overlapping_mappings {
     }
 
     #[ktest]
-    #[should_panic(expected = "cursor virtual address not aligned for mapping")]
+    #[should_panic(expected = "current virtual address is not aligned to the current entry range")]
     fn unaligned_map() {
         const HUGE_PAGE_SIZE: usize = PAGE_SIZE * 512;
 
@@ -508,6 +508,85 @@ mod overlapping_mappings {
     }
 }
 
+mod cursor_range_checks {
+    use super::{test_utils::*, *};
+
+    const HUGE_PAGE_SIZE: usize = PAGE_SIZE * 512;
+    const ALIGNED_CURSOR_RANGE: Range<Vaddr> = 0..PAGE_SIZE;
+    const UNALIGNED_CURSOR_RANGE: Range<Vaddr> = PAGE_SIZE..PAGE_SIZE * 2;
+
+    fn page_table_with_huge_mapping() -> PageTable<TestPtConfig> {
+        let page_table = PageTable::<TestPtConfig>::empty();
+        let prop = PageProperty::new_user(PageFlags::RW, CachePolicy::Writeback);
+        map_untracked(&page_table, 0..HUGE_PAGE_SIZE * 2, 0, prop);
+        page_table
+    }
+
+    #[ktest]
+    fn query_allows_range_outside_cursor() {
+        let page_table = page_table_with_huge_mapping();
+        let preempt_guard = disable_preempt();
+        let cursor = page_table
+            .cursor(&preempt_guard, &UNALIGNED_CURSOR_RANGE)
+            .unwrap();
+
+        assert_eq!(cursor.level(), 2);
+        assert!(matches!(cursor.query(), PteStateRef::Mapped(_)));
+    }
+
+    #[ktest]
+    #[should_panic(expected = "current virtual address range exceeds cursor range")]
+    fn protect_rejects_range_outside_cursor() {
+        let page_table = page_table_with_huge_mapping();
+        let preempt_guard = disable_preempt();
+        let mut cursor = page_table
+            .cursor_mut(&preempt_guard, &ALIGNED_CURSOR_RANGE)
+            .unwrap();
+
+        assert_eq!(cursor.level(), 2);
+        unsafe { cursor.protect(&mut |_| {}) };
+    }
+
+    #[ktest]
+    #[should_panic(expected = "current virtual address range exceeds cursor range")]
+    fn unmap_rejects_range_outside_cursor() {
+        let page_table = page_table_with_huge_mapping();
+        let preempt_guard = disable_preempt();
+        let mut cursor = page_table
+            .cursor_mut(&preempt_guard, &ALIGNED_CURSOR_RANGE)
+            .unwrap();
+
+        assert_eq!(cursor.level(), 2);
+        let _ = unsafe { cursor.unmap() };
+    }
+
+    #[ktest]
+    #[should_panic(expected = "current virtual address is not aligned to the current entry range")]
+    fn protect_rejects_unaligned_current_address() {
+        let page_table = page_table_with_huge_mapping();
+        let preempt_guard = disable_preempt();
+        let mut cursor = page_table
+            .cursor_mut(&preempt_guard, &UNALIGNED_CURSOR_RANGE)
+            .unwrap();
+
+        assert_eq!(cursor.level(), 2);
+        unsafe { cursor.protect(&mut |_| {}) };
+    }
+
+    #[ktest]
+    #[should_panic(expected = "current virtual address is not aligned to the current entry range")]
+    fn unmap_rejects_unaligned_current_address() {
+        let page_table = page_table_with_huge_mapping();
+        let preempt_guard = disable_preempt();
+        let mut cursor = page_table
+            .cursor_mut(&preempt_guard, &UNALIGNED_CURSOR_RANGE)
+            .unwrap();
+
+        assert_eq!(cursor.level(), 2);
+        let _ = unsafe { cursor.unmap() };
+    }
+}
+
 mod navigation {
     use super::{test_utils::*, *};
 
diff --git a/ostd/src/mm/vm_space.rs b/ostd/src/mm/vm_space.rs
index 053d015ab..9832ad217 100644
--- a/ostd/src/mm/vm_space.rs
+++ b/ostd/src/mm/vm_space.rs
@@ -421,7 +421,13 @@ impl<'a> CursorMut<'a> {
     ///
     /// # Panics
     ///
-    /// Panics if the current virtual address is already mapped.
+    /// Panics if:
+    ///  - the current virtual address is not aligned to the current entry
+    ///    range;
+    ///  - the current virtual address range is not fully contained in the
+    ///    cursor range;
+    ///  - the cursor level does not match the frame's mapping level;
+    ///  - the current virtual address is already mapped.
     pub fn map(&mut self, frame: UFrame, prop: PageProperty) {
         let item = VmItem::new_tracked(frame, prop);
 
@@ -508,6 +514,14 @@ impl<'a> CursorMut<'a> {
     /// unmapped range.
     ///
     /// The number of unmapped frames is returned.
+    ///
+    /// # Panics
+    ///
+    /// Panics if:
+    ///  - the current virtual address is not aligned to the current entry
+    ///    range;
+    ///  - the current virtual address range is not fully contained in the
+    ///    cursor range.
     pub fn unmap(&mut self) -> usize {
         // SAFETY:
         // 1. It is safe to unmap memory in the userspace.
@@ -575,6 +589,14 @@ impl<'a> CursorMut<'a> {
     /// cursor, and can be queried via [`Self::cur_va_range`]. Adjust the
     /// level via [`Self::adjust_level`] before protecting to change the
     /// protected range.
+    ///
+    /// # Panics
+    ///
+    /// Panics if:
+    ///  - the current virtual address is not aligned to the current entry
+    ///    range;
+    ///  - the current virtual address range is not fully contained in the
+    ///    cursor range.
     pub fn protect(&mut self, mut op: impl FnMut(&mut PageFlags, &mut CachePolicy)) {
         // SAFETY: It is safe to set `PageFlags` and `CachePolicy` of memory
         // in the userspace.
```

</p>
</details> 

> Could you analyze it further to determine where it comes from?

Yes I admit that these AI generated code is not ideal. Yet if we do the best I think it still adds complexity. I am doubtful if the effort worth if the only problem is more panic cases.

> I don't understand why the `VmSpace` types do not wrap the `Subtree`/`Leaf` types.

Because operations (e.g., `unmap`) of `page_table::Subtree`/`page_table::Leaf` types are unsafe. And `VmSpace` types still need to wrap the `vm_space::Cursor` to access the TLB flusher.

lrh2000 commented:
> I am doubtful if the effort worth if the only problem is more panic cases.

I think the problem is that exposing the cursor level directly leads to rather surprising API surface. I think it's very odd to require the caller to adjust the level manually.

In particular. I think following code is not ideal:
```rust
            while cursor.push_level_if_exists().is_some() {}
            let item = cursor.query();
```
```rust
        while cursor.find_next_unmappable_subtree(range.end).is_some() {
            while cursor.cur_va_range().start < range.start || cursor.cur_va_range().end > range.end
            {
                cursor.adjust_level(cursor.level() - 1);
            }
            num_unmapped += cursor.unmap();
        }
```
```rust
                VmQueriedItem::PageTable => {
                    unreachable!("pushed but still queried a page table")
                }
```
(There may be more, but I just selected some.)

Furthermore, I think we should remove the `VmQueriedItem::None` and `VmQueriedItem::PageTable` variants. `Leaf::item()` shouldn't return them, after all. This is also not done in your prototype. However, I think we should do it so that the benefit is clearer.

> Because operations (e.g., `unmap`) of `page_table::Subtree`/`page_table::Leaf` types are unsafe.

Yes. So why `vm_space::Subtree` wraps `page_table::CursorMut` instead of `page_table::Subtree`? This does not answer my question.

junyang-zh commented:
@lrh2000 

> I think it's very odd to require the caller to adjust the level manually.

Yes I admit that the caller's additional code seems ugly. However in CortenMM the caller indeed need more control and they are wrapped in macros to make it concise. This PR is part of the effort to move forward.

> I think we should remove the VmQueriedItem::None and VmQueriedItem::PageTable variants.

Also this is needed in CortenMM.

I agree that we cannot include CortenMM as the motivation of this standalone PR, I think I can move the last commit to #2615 .

> Yes. So why `vm_space::Subtree` wraps `page_table::CursorMut` instead of `page_table::Subtree`?

In https://github.com/asterinas/asterinas/commit/93352fd7009c8754eb5dff8762ba3aef4ca0aebb `vm_space::Subtree` wraps `vm_space::CursorMut` instead of `page_table::CursorMut`. I believe "And `VmSpace` types still need to wrap the `vm_space::Cursor` to access the TLB flusher." answers your question.

lrh2000 commented:
> Also this is needed in CortenMM.

Why doesn't it work with the `Subtree` APIs? These also allow you to select the level without exposing it directly in the cursor.
```rust
fn query_subtree(&mut self, level: PagingLevel) -> Subtree<'_, 'rcu>;
```

> I agree that we cannot include CortenMM as the motivation of this standalone PR

The real problem is that you never well explained what you need, except from some obscure words like "more control".

For now, the most detailed response from you about what CortenMM needs is the following statement:
>> The motivation to make the cursor level externally visible is to enable CortenMM. Since when storing extra state (VMA pieces actually) in the page table, the caller need to put it in the correct level. Therefore the cursor must allow the user to navigate to specific levels and do operations.

I think `Subtree` APIs work much better for exactly the same purpose. Avoiding `VmQueriedItem::None` and `VmQueriedItem::PageTable` is a real benefit when we're knowing at a non-empty, leaf PTE.

> I believe "And `VmSpace` types still need to wrap the `vm_space::Cursor` to access the TLB flusher." answers your question.

Shouldn't `vm_space::Subtree` wraps `page_table::Subtree`, plus a mutable reference to the TLB flusher of `vm_space::Cursor`? I'm still not sure what I missed.

junyang-zh commented:
> Shouldn't vm_space::Subtree wraps page_table::Subtree, plus a mutable reference to the TLB flusher of vm_space::Cursor?

That, would be two mutable references to two fields in one object, which would be rejected by the compiler?

lrh2000 commented:
> > Shouldn't vm_space::Subtree wraps page_table::Subtree, plus a mutable reference to the TLB flusher of vm_space::Cursor?
> 
> That, would be two mutable references to two fields in one object, which would be rejected by the compiler?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=1b379df1f1c334dcbe5b3f69f7461185

junyang-zh commented:
Sorry for the delay.

> I think the problem is that exposing the cursor level directly leads to rather surprising API surface. I think it's very odd to require the caller to adjust the level manually.
>
> In particular. I think following code is not ideal:

Let me clarify my position first. The ideal Cursor API I want is actually a minimal, self-contained set of operations for arbitrary complicated operations. This PR is still not ideal in this sense. Actually, `find_next` should be removed from OSTD, and we should let the caller implement such logic. Complexity on the caller side is a minor goal. We can always add a layer of helpers for it. However, ambiguity on the caller side is a bigger problem.

The most problematic ambiguity so far is whether the cursor will automatically modify the page table upon queries. E.g., if `query_leaf` is called on a huge page or an empty upper-level entry, would it split the page/add new child page tables? What if `find_next_unmappable_subtree` is called on a huge page with the end exceeding the `end` but the start ahead of `end`? In your ideal case, it seems that automatic modifications are needed to make the caller site simple. But I want such a caller site to explicitly say what modifications are needed. I don't think the subtree/leaf API helps in this case and adds more complexity if we do not want automatic modifications.

Furthermore, I don't think there is any way to reduce the net added lines to less than 100 to implement the subtree/leaf API, even if your suggestions (avoiding `VmQueriedItem::None`, adding `Leaf::to_subtree()`, merging definitions, ...) are applied.

junyang-zh commented:
I added some wrappers in the kernel. Now it looks like:

```
            let item = cursor.to_leaf().query();
```

```
        while cursor.find_next_unmappable_subtree(range.end).is_some() {
            cursor.split_if_map_exceeds_range(&range);
            num_unmapped += cursor.unmap();
        }
```
